uitools/widgets/Esri/ArcGISRuntime/Toolkit/FloorFilter.cpp

```diff
@@ -0,0 +1,313 @@
+     \brief Given a \a modelIndex, for \a model,
+     extracts the `T*` pointer from its \c userRole.
+     */
+    template <typename T>
+    T* itemForIndex(QAbstractItemModel* model, QModelIndex modelIndex)
+    {
+      if (!model)
+        return nullptr;
+
+      if (modelIndex.isValid())
+      {
+        return qvariant_cast<T*>(model->data(modelIndex, Qt::UserRole));
+      }
+      return nullptr;
+    }
+
+    /*!
+     \internal
+     \brief Given some modelId in a given model, returns that items index in the model.
+            O(n) search time.
+     */
+    template <typename T>
+    QModelIndex indexForId(QAbstractItemModel* model, QString id)
+    {
+      const int rowCount = model->rowCount();
+      for (int i = 0; i < rowCount; ++i)
+      {
+        const auto index = model->index(i, 0);
+        const auto item = itemForIndex<T>(model, index);
+        if (item && item->modelId() == id)
+          return index;
+      }
+      return QModelIndex{};
+    }
+
+  /*!
+    \internal
+    \brief Wrap the \a sourceModel with a \l QSortFilterProxyModel that sorts and filters based on the \a filterField textEdited event.
+   */
+    QSortFilterProxyModel* wrapFilterModel(QAbstractItemModel* sourceModel, QLineEdit* filterField, FloorFilter* parent)
+    {
+      auto model = new QSortFilterProxyModel(parent);
+      model->setSourceModel(sourceModel);
+      model->setSortCaseSensitivity(Qt::CaseInsensitive);
+      model->setFilterCaseSensitivity(Qt::CaseInsensitive);
+
+      QObject::connect(filterField, &QLineEdit::textEdited, parent, [model](const QString& text)
+                       {
+                         model->setFilterFixedString(text);
+                       });
+      return model;
+    }
+
+    /*!
+     \internal
+     \brief Helper struct that holds a reference to sme `b`, sets `b` to `v` on construction,
+     and then sets `b` back to its initial value on destruction.
+     */
+    template <typename T>
+    struct PushValue
+    {
+      PushValue(T& b, T v) :
+        m_tracked(b),
+        m_state{std::move(m_tracked)}
+      {
+        m_tracked = std::move(v);
+      }
+
+      ~PushValue()
+      {
+        m_tracked = std::move(m_state);
+      }
+
+      T& m_tracked;
+      T m_state;
+    };
+
+    /*!
+     \internal
+     \brief Creates a `PushValue`.
+     */
+    template <typename T>
+    PushValue<T> push_value(T& t, T v)
+    {
+      return PushValue<T>(t, v);
+    }
+  }
+
+  /*!
+    \inmodule EsriArcGISRuntimeToolkit
+    \class Esri::ArcGISRuntime::Toolkit::FloorFilter
+    \brief Allows to display and filter the available floor aware layers in the current \c GeoModel.
+    The FloorFilter allows the interaction with the available floor aware layers. A user can select from a list of sites which presents
+    their facilities. Once a facility is chosen, it is possible to toggle between its levels which will show them on the \c GeoView.
+    2D maps and 3D scenes are supported.
```

Comment:
Here is the suggestion I made in the QML doc to add a bit more info. Feel free to rework

```suggestion
  \brief Displays and filters the available floor aware layers in the current \c GeoModel.

  This component supports manual selection and text-based filtering for sites and facilities.
  Once a site and facility is selected, levels can be selected and the GeoView will be updated.
  
  In addition to the browsing workflow where sites and facilities are manually selected, automatic facility 
  selection based on the GeoView's current viewpoint. This mode is useful when you are exploring the
  GeoView but do not know the name of the site or facility.
  
  The user interface is driven from the FloorAware data that is available in the GeoModel's FloorManager. 
  
  2D maps and 3D scenes are supported.
```